engineioxide/src/payload.rs

```diff
@@ -0,0 +1,140 @@
+use std::{io::BufRead, vec};
+
+use crate::{service::ProtocolVersion, errors::Error};
+
+pub const PACKET_SEPARATOR: u8 = b'\x1e';
+
+/// A payload is a series of encoded packets tied together.
+/// How packets are tied together depends on the protocol.
+pub struct Payload<R: BufRead> {
+    reader: R,
+    buffer: Vec<u8>,
+    #[allow(dead_code)]
+    protocol: ProtocolVersion,
+}
+
+type Item = Result<String, Error>;
+
+impl<R: BufRead> Payload<R> {
+    pub fn new(protocol: ProtocolVersion, data: R) -> Self {
+        Payload {
+            reader: data,
+            buffer: vec![],
+            protocol,
+        }
+    }
+
+    #[cfg(feature = "v3")]
+    fn next_v3(&mut self) -> Option<Item> {
+        match self.reader.read_until(b':', &mut self.buffer) {
+            Ok(bytes_read) => (bytes_read > 0).then(|| {
+                if self.buffer.ends_with(&[b':']) {
+                    self.buffer.pop();
```

Comment:
You can just parse to str rather than String, it avoids reallocating.
```suggestion
  let length = match std::str::from_utf8(&self.buffer) {
```

Reply:
Even for the whole Item returned it is possible to return a `str` rather than a `String`, but maybe we could implement than in another PR

Reply:
Changed it for length only for now. Agreed, we could make it return a str packet altogether but it appears to be a larger refactor so it's fine with me for someone to finish that in another PR.